staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -0,0 +1,207 @@
+// resourceVersion.
+type watchCacheInterval struct {
+	// startIndex denotes the starting point
+	// of interval being considered, the index
+	// value is the index in the actual watch
+	// cache, however this should be used modulo
+	// the capacity of the cache whenever being
+	// used to index the circular buffer.
+	startIndex int
+
+	// endIndex denotes the end of the interval
+	// being considered, the index value is the
+	// index in the actual watch cache, however
+	// this should be used modulo the capacity
+	// of the cache whenever being used to index
+	// the circular buffer.
+	endIndex int
+
+	// indexer takes an index and returns the corresponding
+	// watchCacheEvent from an underlying watch cache.
+	indexer func(int) *watchCacheEvent
+
+	// intervalBuffer is a buffer of fixed size to hold the
+	// watchCacheEvents that this interval returns on a call to
+	// Next(). This exists mainly to reduce acquiring the lock
+	// on each invocation of Next().
+	intervalBuffer *watchCacheIntervalBuffer
+
+	// lock effectively guards the valid field.
+	lock sync.Locker
+
+	// valid indicates whether the cache interval
+	// being considered (b/w startIndex and endIndex)
+	// is still valid or not.
+	valid bool
+}
+
+type attrFunc func(runtime.Object) (labels.Set, fields.Set, error)
+
+func newCacheInterval(startIndex, endIndex int, locker sync.Locker) *watchCacheInterval {
+	return &watchCacheInterval{
+		startIndex:     startIndex,
+		endIndex:       endIndex,
+		intervalBuffer: &watchCacheIntervalBuffer{buffer: make([]*watchCacheEvent, bufferSize)},
+		lock:           locker,
+	}
+}
+
+// newCacheIntervalFromStore is meant to handle the case of rv=0, such that the events
+// returned by Next() need to be events from a List() done on the underlying store of
+// the watch cache.
+func newCacheIntervalFromStore(store cache.Indexer, getAttrsFunc attrFunc, locker sync.Locker) (*watchCacheInterval, error) {
+	ciBuffer := &watchCacheIntervalBuffer{}
+	allItems := store.List()
+	ciBuffer.buffer = make([]*watchCacheEvent, len(allItems))
+	for i, item := range allItems {
+		elem, ok := item.(*storeElement)
+		if !ok {
+			return nil, fmt.Errorf("not a storeElement: %v", elem)
+		}
+		objLabels, objFields, err := getAttrsFunc(elem.Object)
+		if err != nil {
+			return nil, err
+		}
+		ciBuffer.buffer[i] = &watchCacheEvent{
+			Type:            watch.Added,
+			Object:          elem.Object,
+			ObjLabels:       objLabels,
+			ObjFields:       objFields,
+			Key:             elem.Key,
+			ResourceVersion: 0,
+		}
+	}
+	ci := &watchCacheInterval{intervalBuffer: ciBuffer, lock: locker}
+	// simulate that we already have all the events we're looking for.
+	ci.startIndex = ci.endIndex
+
+	return ci, nil
+}
+
+// registerIndexerFunc registers an indexer that is used to retrieve an event from the
+// underlying watch cache given an index.
+func (wci *watchCacheInterval) registerIndexerFunc(indexer func(int) *watchCacheEvent) {
```

Comment:
Can you remove this function and just make this function a parameter of the newCacheInterval
[it shouldn't be mutable anyway - so let's just avoid having this function]